src/Features/Core/Portable/CodeRefactoringHelpers.cs

```diff
@@ -0,0 +1,120 @@
+            // with the selection starting somewhere between the full node-line span, and the start
+            // of the node as well as ending after the end of the node and the end of the line the
+            // node ends on.
+            var expandedSpan = await GetExpandedNodeSpan(document, node, cancellationToken).ConfigureAwait(false);
+            if ((selection.Start >= expandedSpan.Start && selection.Start <= node.SpanStart) &&
+                (selection.End >= node.Span.End && selection.End <= expandedSpan.End))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Determines if a refactoring should be offered for a given node given the specified
+        /// position in a document. The refactoring is offered if the position is somewhere on the
+        /// span containing the node, and not in any of the specified <paramref name="holes"/>
+        /// to avoid.
+        ///
+        /// Note: this function considers the span containing the node to start at the earliest
+        /// preceding whitespace before the node's <see cref="SyntaxNode.SpanStart"/> and up through
+        /// the last whitespace following the <see cref="SyntaxNode.Span"/>’s <see cref="TextSpan.End"/>.
+        /// In other words, if the position is within the <see cref="SyntaxNode.FullSpan"/>
+        /// of the node but on a line preceding <see cref="SyntaxNode.SpanStart"/> then it is not a
+        /// valid position. This prevents refactorings from showing up on comments or whitespace on
+        /// preceding lines.
+        ///
+        /// For the <paramref name="holes"/>, the position is considered invalid if it is *within*
+        /// the hole, not if it is touching the hole edges.
+        /// </summary>
+        public static async Task<bool> RefactoringPositionIsValidAsync(
+            Document document,
+            int position,
+            SyntaxNode node,
+            ImmutableArray<SyntaxNode> holes,
+            CancellationToken cancellationToken)
+        {
+            var expandedSpan = await GetExpandedNodeSpan(document, node, cancellationToken).ConfigureAwait(false);
+            if (!expandedSpan.IntersectsWith(position))
+            {
+                // Position isn’t within the node’s widened span. Definitely not valid here.
+                return false;
+            }
+
+            foreach (var hole in holes)
+            {
+                if (position > hole.Span.Start && position < hole.Span.End)
+                {
+                    // Position was in one of the holes. Not valid here.
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private static async Task<TextSpan> GetExpandedNodeSpan(
+            Document document,
+            SyntaxNode node,
+            CancellationToken cancellationToken)
+        {
```

Comment:
did any of these helpers change since when i wrote and gave to you?    Just want to know if you needed to update any of hte logic, or if it worked as is.  Thanks!

Reply:
@CyrusNajmabadi Your original code is at https://gitter.im/jnm2?at=5ba04beff4bd1056ac7aabd3 (link only works for you).

Diff of my changes to it:

<details>

```diff
@@ -27,7 +27,7 @@ internal static class CodeRefactoringHelpers
                     document, selection.Start, node, holes, cancellationToken).ConfigureAwait(false);
             }

-            // If we have selection it needs to be selecting at least the full node.  We are lenient
+            // If we have selection, it needs to be selecting at least the full node. We are lenient
             // with the selection starting somewhere between the full node-line span, and the start
             // of the node as well as ending after the end of the node and the end of the line the
             // node ends on.
@@ -43,29 +43,32 @@ internal static class CodeRefactoringHelpers

         /// <summary>
         /// Determines if a refactoring should be offered for a given node given the specified
-        /// position in a document.  The refactoring is offered if the position is somewhere on the
+        /// position in a document. The refactoring is offered if the position is somewhere on the
         /// span containing the node, and not in any of the specified <paramref name="holes"/>
         /// to avoid.
         ///
-        /// Note: this function consider the span containing the node to start at the earliest
+        /// Note: this function considers the span containing the node to start at the earliest
         /// preceding whitespace before the node's <see cref="SyntaxNode.SpanStart"/> and up through
-        /// the last whitespace following the <cref="SyntaxNode.Span"/>'s <see cref="TextSpan.End"/>.
+        /// the last whitespace following the <see cref="SyntaxNode.Span"/>’s <see cref="TextSpan.End"/>.
         /// In other words, if the position is within the <see cref="SyntaxNode.FullSpan"/>
-        /// of the node, but on a line preceding <see cref="SyntaxNode.SpanStart"/> then it is not a
-        /// valid position.  This prevents refactorings from showing up on comments or whitespace on
+        /// of the node but on a line preceding <see cref="SyntaxNode.SpanStart"/> then it is not a
+        /// valid position. This prevents refactorings from showing up on comments or whitespace on
         /// preceding lines.
         ///
         /// For the <paramref name="holes"/>, the position is considered invalid if it is *within*
         /// the hole, not if it is touching the hole edges.
         /// </summary>
         public static async Task<bool> RefactoringPositionIsValidAsync(
-            Document document, int position,
-            SyntaxNode node, ImmutableArray<SyntaxNode> holes, CancellationToken cancellationToken)
+            Document document,
+            int position,
+            SyntaxNode node,
+            ImmutableArray<SyntaxNode> holes,
+            CancellationToken cancellationToken)
         {
             var expandedSpan = await GetExpandedNodeSpan(document, node, cancellationToken).ConfigureAwait(false);
-            if (expandedSpan.IntersectsWith(position))
+            if (!expandedSpan.IntersectsWith(position))
             {
-                // position isn't within the node's widened span is on.  Definitely not valid here.
+                // Position isn’t within the node’s widened span. Definitely not valid here.
                 return false;
             }

@@ -73,23 +76,29 @@ internal static class CodeRefactoringHelpers
             {
                 if (position > hole.Span.Start && position < hole.Span.End)
                 {
-                    // Position was in one of the holes.  Not valid here.
+                    // Position was in one of the holes. Not valid here.
                     return false;
                 }
             }

-            // this position is good.
             return true;
         }

-
         private static async Task<TextSpan> GetExpandedNodeSpan(
-            Document document, SyntaxNode node, CancellationToken cancellationToken)
+            Document document,
+            SyntaxNode node,
+            CancellationToken cancellationToken)
         {
-            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
             var sourceText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);

             var nodeStartLine = sourceText.Lines.GetLineFromPosition(node.SpanStart);
+
+            // Enable vertical selections that catch the previous line break and perhaps some whitespace.
+            if (nodeStartLine.LineNumber != 0)
+            {
+                nodeStartLine = sourceText.Lines[nodeStartLine.LineNumber - 1];
+            }
+
             var nodeEndLine = sourceText.Lines.GetLineFromPosition(node.Span.End);

             var start = node.SpanStart;
```

</details>